includes/admin/class-wc-rest-payments-reader-controller.php

```diff
@@ -62,4 +100,69 @@ public function get_summary( $request ) {
 		return rest_ensure_response( $summary );
 	}
 
+	/**
+	 * Renders HTML for a print receipt
+	 *
+	 * @param WP_REST_Request $request Full data about the request.
+	 * @return WP_HTTP_RESPONSE|WP_Error
+	 */
+	public function generate_print_receipt( $request ) {
+		try {
+			$payment_intent = $this->api_client->get_intent( $request->get_param( 'payment_id' ) );
+
+			if ( 'succeeded' !== $payment_intent->get_status() ) {
+				return rest_ensure_response( new WP_Error( 'generate_print_receipt_error', __( 'Invalid payment', 'woocommerce-payments' ), [ 'status' => 500 ] ) );
+			}
+
+			$charge = $this->api_client->get_charge( $payment_intent->get_charge_id() );
+		} catch ( API_Exception $e ) {
+			return rest_ensure_response( new WP_Error( 'generate_print_receipt_error', $e->getMessage(), [ 'status' => $e->get_http_code() ] ) );
+		}
+
+		$order = wc_get_order( $charge['order']['number'] );
+
+		if ( ! $order ) {
+			return rest_ensure_response( new WP_Error( 'generate_print_receipt_error', __( 'Order not found', 'woocommerce-payments' ), [ 'status' => 500 ] ) );
+		}
+
+		try {
+			$settings = [
+				'business_name' => $this->wcpay_gateway->get_option( 'account_business_name' ),
+				'support_info'  => [
+					'address' => $this->wcpay_gateway->get_option( 'account_business_support_address' ),
+					'phone'   => $this->wcpay_gateway->get_option( 'account_business_support_phone' ),
+					'email'   => $this->wcpay_gateway->get_option( 'account_business_support_email' ),
+				],
+			];
+		} catch ( Exception $e ) {
+			return rest_ensure_response( new WP_Error( 'generate_print_receipt_error', __( 'There was a problem retrieving merchant settings.', 'woocommerce-payments' ), [ 'status' => 500 ] ) );
+		}
+
+		/**
+		 * WP_REST_Server will convert the response data to JSON prior to output it.
+		 * Using this filter to prevent it, and output the data from WP_HTTP_Responde instead.
+		 */
+		add_filter(
+			'rest_pre_serve_request',
+			function ( bool $served, WP_HTTP_Response $response ) : bool {
+				$served = true;
+				echo $response->get_data(); // @codingStandardsIgnoreLine
+				return $served;
+			},
+			10,
+			2
+		);
```

Comment:
I think this block should be moved after the `try ... catch`:
1. The filter is more related to `return new WP_HTTP_Response` than to `$this->receipts_service->get_receipt_markup` so it makes sense that it's defined right before we're returning response.
2. We'll try to echo out `WP_Error` in case if `receipts_service` throws exception. Testing locally returns `Notice: Array to string conversion`

Reply:
Thanks for checking it @KarlisJ 
>I think this block should be moved after the try ... catch

Makes sense I'll move it near the return 👍 
> We'll try to echo out WP_Error in case if receipts_service throws exception. Testing locally returns Notice: Array to string conversion

Yes good catch, I just noticed this problem. I am going to echo the data only if the response is successful, otherwise the callback will return false, causing the response (error) to be handled by WP_REST_Server

Reply:
AFAIU, no need to check for a successful response if you define the filter right before returning.

Reply:
>AFAIU, no need to check for a successful response if you define the filter right

You are absolutely right